### ISSUE TITLE
Propagate container configuration from base image?

### ISSUE THREAD

**Issue by coollog** [GoogleContainerTools/jib#595]
Thumbs up if this is useful for you.

**Comment by chanseokoh** [comment]
Hmm... if the images built by docker in the normal way don't inherit the container configuration, I think we don't need to do this. OTOH, if they do, this is something we can consider.

**Comment by coollog** [comment]
Yes, I believe Dockerfile `FROM` uses the container configuration from the base image. Any instructions afterwards are extensions on that base container configuration.

**Comment by essh** [comment]
This is useful for me to pick up on the Env from a base image (specifically `PATH` and `JAVA_HOME`). As already mentioned these were previously coming along when using a DockerFile `FROM` and would save having to specify them when launching a container built with jib.

**Comment by coollog** [comment]
I think for this feature, we can start with propagating environment variables since that seems to be the most useful case.

**Comment by TadCordle** [comment]
Adjusting the milestone since we were only planning on propagating environment variables for this next release.

**Comment by chanseokoh** [comment]
For those who are following this issue, environment variables from the base image are not being propagated into the target image.

**Comment by chanseokoh** [comment]
> environment variables from the base image are not being propagated

Sorry, typo. Environment variables from the base image are **now** being propagated

**Comment by chanseokoh** [comment]
Turns out we also need to propagate `workingDir` to make use of distroless-jetty as a base image for #431. Will work on this.

**Comment by chanseokoh** [comment]
Also `ExposedPorts` in the container configuration.

**Comment by chanseokoh** [comment]
#924: `Entrypoint` and `CMD`

**Comment by GuustavoPaiva** [comment]
@chanseokoh I like using the base image ports if none is specified. 

Also, I think volumes must be inherited as well. After the PR for supporting volume configuration (#1215) is accepted we can do it.

**Comment by chanseokoh** [comment]
@GuustavoPaiva yeah, we actually do need to inherit the ports and the volumes, which docker does with `FROM` in `Dockerfile`.

**Comment by coollog** [comment]
We have two left: volumes and healthcheck

Updated original post with a TODO list.

**Comment by chanseokoh** [comment]
I think we haven't handled ports either. Updated the TODO list.

**Comment by chanseokoh** [comment]
Just in case, I am documenting here that there is one subtle to be careful about when enabling inheriting ports, volumes, etc. In `BuildImageSteps`, the code currently looks like

```java
      // inherit properties
      imageBuilder
          .addEnvironment(baseImage.getEnvironment())
          .addLabels(baseImage.getLabels())
          .setWorkingDirectory(baseImage.getWorkingDirectory());
      ...
      // set properties when explicitly given
      if (containerConfiguration != null) {
        imageBuilder
            .addEnvironment(containerConfiguration.getEnvironmentMap())
            ...
            .setExposedPorts(containerConfiguration.getExposedPorts())
            .setVolumes(containerConfiguration.getVolumes())
            ...
      }
```

To inherit volumes for example, one would start with

```java
      // inherit properties
      imageBuilder
          ...
          .setWorkingDirectory(baseImage.getWorkingDirectory())
          .setVolumes(baseImage.getVolumes());  // inherit volumes, YEAH!
      ...
      // set properties when explicitly given
      if (containerConfiguration != null) {
        ...
      }
```

In addition to this, we should update the code to selectively override properties:

```java
      // inherit properties
      imageBuilder
          ...
          .setVolumes(baseImage.getVolumes());  // inherit volumes, YEAH!

      // set properties when explicitly given
      if (containerConfiguration != null) {
        ...
        // should override only when actually given
        if (containerConfiguration.getExposedPorts() != null) {
            imageBuilder.setExposedPorts(containerConfiguration.getExposedPorts());
        }
        // should override only when actually given
        if (containerConfiguration.getVolumes() != null) {
            imageBuilder.setVolumes(containerConfiguration.getVolumes());
        }
      }
```

**Comment by coollog** [comment]
I think we should be adding the volumes/ports rather than replacing them since that's the behavior of Dockerfile builds as well. So, having port `1234` on the base image and setting port `5678` to expose would result in an image with ports `1234` and `5678` exposed.

**Comment by coollog** [comment]
@essh @cwensel @thekguy @sunnychanwork @joelhandwell @jcrossley3 @Tanmayshetty All of these configurations are propagated from the base image now in version `0.10.1`!
